lib/charms/loki_k8s/v0/loki_push_api.py

```diff
@@ -907,7 +1062,67 @@ def alerts(self) -> dict:
         return alerts
 
 
-class LokiPushApiConsumer(RelationManagerBase):
+class ConsumerBase(RelationManagerBase):
+    """Consumer's base class."""
+
+    def __init__(
+        self,
+        charm: CharmBase,
+        relation_name: str = DEFAULT_RELATION_NAME,
+        alert_rules_path: str = DEFAULT_ALERT_RULES_RELATIVE_PATH,
+        allow_free_standing_rules: bool = False,
+    ):
+        super().__init__(charm, relation_name)
+        self._charm = charm
+        self._relation_name = relation_name
+        self.allow_free_standing_rules = allow_free_standing_rules
+        self._alert_rules_path = _resolve_dir_against_charm_path(charm, alert_rules_path)
+        self.topology = JujuTopology.from_charm(charm)
+
+    def _handle_alert_rules(self, relation):
+        if self._charm.unit.is_leader():
+            alert_groups, invalid_files = load_alert_rules_from_dir(
+                self._alert_rules_path,
+                self.topology,
+                recursive=False,
+                allow_free_standing=self.allow_free_standing_rules,
+            )
+            alert_rules_error_message = self._check_alert_rules(alert_groups, invalid_files)
+
+            if alert_rules_error_message:
+                self.on.loki_push_api_alert_rules_error.emit(alert_rules_error_message)
+
+            relation.data[self._charm.app]["metadata"] = json.dumps(self.topology.as_dict())
+            relation.data[self._charm.app]["alert_rules"] = json.dumps({"groups": alert_groups})
+
+    def _check_alert_rules(self, alert_groups, invalid_files) -> str:
+        """Check alert rules.
+
+        Args:
+            alert_groups: a list of prometheus alert rule groups.
+            invalid_files: a list of invalid rules files.
+
+        Returns:
+            A string with the validation message. The message is not empty whether there are
+            invalid alert rules files or there are no alert rules groups.
+        """
+        message = ""
+
+        if invalid_files:
+            must_contain = ["'alert'", "'expr'"]
+            if self.allow_free_standing_rules:
+                must_contain.append("'%%juju_topology%%'")
```

Comment:
Looks familiar but I'm confused: free standing means no juju topology stub, no? i.e. is there a `not` missing?
```suggestion
            if not self.allow_free_standing_rules:
                must_contain.append("'%%juju_topology%%'")
```

Reply:
Wasn't it you who added this part? ;-)

In other parts of the code is `if not ...`

What do you think @dstathis ?

Reply:
Intuitively `not` seems correct but I think I'm not the expert here.